File: src/infra/eventbus.ts
/**
 * TiaLynn 事件总线（前端）。
 *
 * 模块间通信必须经此（不允许跨域 import 内部函数）。
 * 事件命名约定：`<domain>:<verb>_<noun>`，全小写下划线。
 *
 * Rust ↔ 前端 通信仍用 Tauri `emit` / `listen`。
 */
import mitt, { type Emitter } from 'mitt'
import type { EmotionId } from '@/brain/types/soul'

// mitt 要求 Record<EventType, unknown> —— 用 type 别名而非 interface
// （interface 在 TS 中不自动满足索引签名）
export type EventMap = {
  // ---------- Infra ----------
  'infra:config_loaded': void
  'infra:config_changed': { keys: string[] }
  'infra:soul_reloaded': void
  'infra:hotkey_pressed': { key: string }
  'infra:app_quitting': void

  // ---------- Brain ----------
  'brain:chat_input': { text: string }
  'brain:reply_token': { stream_id: string; delta: string }
  'brain:reply_end': {
    stream_id: string
    full_text: string
    emotion?: EmotionId
    intensity?: number
  }
  'brain:emotion_changed': { emotion: EmotionId; intensity: number }
  'brain:thinking_start': { stream_id: string }
  'brain:thinking_end': { stream_id: string }

  // ---------- Avatar ----------
  'avatar:model_loaded': { model_path: string; cubism: string }
  'avatar:model_load_failed': { reason: string }
  'avatar:mouse_inside': { inside: boolean }
  'avatar:dragged': { x: number; y: number }

  // ---------- Presence ----------
  'presence:tts_start': { audio_path: string; emotion: EmotionId | string }
  'presence:tts_end': void
  'presence:tts_error': { reason: string }
  'presence:stt_start': void
  'presence:stt_result': { text: string }
  'presence:stt_error': { reason: string }
  'presence:proactive_speak': { hint: string } // M5

  // ---------- Hands （M4） ----------
  'hands:tool_request': { tool: string; args: Record<string, unknown> }
  'hands:tool_result': { tool: string; ok: boolean; output?: unknown; error?: string }
  'hands:approval_required': { tool: string; args: Record<string, unknown>; id: string }
  'hands:approval_decision': { id: string; approved: boolean }
}

export const bus: Emitter<EventMap> = mitt<EventMap>()

// 便利方法：兼容旧用法（含 payload-less 事件）
export function emit<K extends keyof EventMap>(
  type: K,
  ...args: EventMap[K] extends void ? [] : [EventMap[K]]
): void {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(bus.emit as any)(type, args[0])
}

// 开发模式：把所有事件打到 console，方便调试
if (import.meta.env.DEV) {
  bus.on('*', (type, e) => {
    console.debug('[bus]', type, e ?? '')
  })
}

File: src-tauri/src/infra/eventbus.rs
//! Rust 端事件名常量。
//!
//! Rust ↔ 前端 跨进程通信用 Tauri 的 emit/listen，事件名集中定义在这里
//! 避免代码里散落字符串字面量。

#[allow(dead_code)]
pub mod ev {
    // Infra
    pub const INFRA_CONFIG_LOADED: &str = "infra:config_loaded";
    pub const INFRA_CONFIG_CHANGED: &str = "infra:config_changed";
    pub const INFRA_SOUL_RELOADED: &str = "infra:soul_reloaded";
    pub const INFRA_HOTKEY: &str = "infra:hotkey_pressed";

    // Brain
    pub const BRAIN_CHAT_INPUT: &str = "brain:chat_input";
    pub const BRAIN_REPLY_TOKEN: &str = "brain:reply_token";
    pub const BRAIN_REPLY_END: &str = "brain:reply_end";
    pub const BRAIN_EMOTION_CHANGED: &str = "brain:emotion_changed";

    // Avatar
    pub const AVATAR_MOUSE_GLOBAL: &str = "avatar:mouse_global";
    pub const AVATAR_MODEL_LOADED: &str = "avatar:model_loaded";

    // Presence
    pub const PRESENCE_TTS_START: &str = "presence:tts_start";
    pub const PRESENCE_TTS_END: &str = "presence:tts_end";
    pub const PRESENCE_STT_START: &str = "presence:stt_start";
    pub const PRESENCE_STT_RESULT: &str = "presence:stt_result";
    pub const PRESENCE_STT_ERROR: &str = "presence:stt_error";
    pub const PRESENCE_STT_TRANSCRIBING: &str = "presence:stt_transcribing";

    // Hands (M4)
    pub const HANDS_TOOL_REQUEST: &str = "hands:tool_request";
    pub const HANDS_TOOL_RESULT: &str = "hands:tool_result";
    pub const HANDS_APPROVAL_REQUIRED: &str = "hands:approval_required";

    // Sidecar
    pub const SIDECAR_INSTALL_LOG: &str = "sidecar:install_log";
    pub const SIDECAR_INSTALL_DONE: &str = "sidecar:install_done";
}
